Source/carthage/Extensions.swift

```diff
@@ -58,71 +57,112 @@ public struct CommandError {
+			case let .Error(err):
+				error?(err)
+			case .Completed:
+				completed?()
+			case .Interrupted:
+				interrupted?()
+			}
 		}
 	}
 }
 
-/// Transforms the error type in a Result.
-internal func mapError<T, E, F>(result: Result<T, E>, transform: E -> F) -> Result<T, F> {
-	switch result {
-	case let .Success(value):
-		return .Success(value)
+extension SignalProducerType {
+
+	/// Bring back the `start` overload. The `startNext` or pattern matching
+	/// on `start(Event)` is annoying in practice and more verbose. This is also
+	/// likely to change in a later RAC 4 alpha.
+	internal func start(next next: (T -> ())? = nil, error: (E -> ())? = nil, completed: (() -> ())? = nil, interrupted: (() -> ())? = nil) -> Disposable? {
+		return self.start { (event: Event<T, E>) in
+			switch event {
+			case let .Next(value):
+				next?(value)
+			case let .Error(err):
+				error?(err)
+			case .Completed:
+				completed?()
+			case .Interrupted:
+				interrupted?()
+			}
+		}
+	}
+}
 
-	case let .Failure(error):
-		return .Failure(Box(transform(error.value)))
+extension SignalType where E == CarthageError {
+	/// Promotes CarthageErrors into CommandErrors.
+	internal func promoteErrors() -> Signal<T, CommandError> {
+		return signal.mapError { (error: CarthageError) -> CommandError in
+			let commandantError = CommandantError.CommandError(error)
+			return CommandError(commandantError)
+		}
 	}
 }
 
-/// Promotes CarthageErrors into CommandErrors.
-internal func promoteErrors<T>(signal: Signal<T, CarthageError>) -> Signal<T, CommandError> {
-	return signal |> mapError { (error: CarthageError) -> CommandError in
-		let commandantError = CommandantError.CommandError(Box(error))
-		return CommandError(commandantError)
+extension SignalProducerType where E == CarthageError {
+	/// Promotes CarthageErrors into CommandErrors.
+	internal func promoteErrors() -> SignalProducer<T, CommandError> {
+		return lift { $0.promoteErrors() }
 	}
 }
 
 /// Lifts the Result of options parsing into a SignalProducer.
 internal func producerWithOptions<T>(result: Result<T, CommandantError<CarthageError>>) -> SignalProducer<T, CommandError> {
-	let mappedResult = mapError(result) { CommandError($0) }
+	let mappedResult = result.mapError { CommandError($0) }
 	return SignalProducer(result: mappedResult)
 }
 
-/// Waits on a SignalProducer that implements the behavior of a CommandType.
-internal func waitOnCommand<T>(producer: SignalProducer<T, CommandError>) -> Result<(), CommandantError<CarthageError>> {
-	let result = producer
-		|> then(SignalProducer<(), CommandError>.empty)
-		|> wait
-	
-	TaskDescription.waitForAllTaskTermination()
-	return mapError(result) { $0.error }
+extension SignalProducerType where E == CommandError {
+	/// Waits on a SignalProducer that implements the behavior of a CommandType.
+	internal func waitOnCommand() -> Result<(), CommandantError<CarthageError>> {
+		let result = producer
+			.then(SignalProducer<(), CommandError>.empty)
```

Comment:
Unfortunately that does not compile (compiler can't infer the correct type). :disappointed: